### PR TITLE
Windows MF Camera Node

### PR DESCRIPTION
**Pull request by Sudhaushu**
MF based camera node. 
Resolution configs added. - dynamic change in response to change in cameraInfo via setCameraInfo

TBD-
Compressed stream/topic if Mjpeg is available from camera - not working yet.
Tests need to be added

### CODE REVIEW

File: src/wincameranodelet.cpp [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                    }
+                    waitForFinish.unlock();
+                }
+            };
+
+            auto handler1 = [&](winrt::hresult_error ex, winrt::hstring msg, IMFSample* pSample)
+            {
+                if (pSample)
+                {
+                    std::cout << "Received SAmple handller1\n";
```
Comment (ooeygui) [inline src/wincameranodelet.cpp]: Please convert to a ROS_INFO.

File: src/wincameranodelet.cpp [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+                    videoSourcePath = winrt::to_string(ros_win_camera::WindowsMFCapture::EnumerateCameraLinks(false).First().Current());
+                }
+            }
+            std::shared_ptr<camera_info_manager::CameraInfoManager> spCameraInfoManager = std::make_shared<camera_info_manager::CameraInfoManager>(privateNode, "frame_id");
+            WinRosPublisherImageRaw rawPublisher(privateNode, "image_raw", PUBLISHER_BUFFER_SIZE, frame_id, spCameraInfoManager.get());
+            auto handler = [&](winrt::hresult_error ex, winrt::hstring msg, IMFSample* pSample)
+            {
+                if (pSample)
+                {
+                    //castd::cout << "Received SAmple\n";
```
Comment (ooeygui) [inline src/wincameranodelet.cpp]: Remove comment.

File: src/wincameranodelet.cpp [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        }
+        ~WinCameraNodelet()
+        {
+            waitForFinish.lock();
+        }
+
+    private:
+        virtual void onInit()//int main(int argc, char** argv)
+        {
+            //ros::init(argc, argv, "win_camera");
```
Comment (ooeygui) [inline src/wincameranodelet.cpp]: Remove commented code

File: include/winrospublisher.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
+#include <mfapi.h>
+#include <winrt\base.h>
```
Comment (NaveenThum) [inline include/winrospublisher.h:3]: Copyrights?

File: include/wincapture.h [hunk 5]
```diff
@@ -0,0 +1,99 @@
+#include <mfapi.h>
+#include <mfidl.h>
```
Comment (NaveenThum) [inline include/wincapture.h:3]: Copyrights?

File: src/wincameranode.cpp [hunk 6]
```diff
@@ -0,0 +1,140 @@
+#include "wincapture.h"
+#include "winrospublisher.h"
```
Comment (NaveenThum) [inline src/wincameranode.cpp:3]: Copyrights?

File: src/wincameranodelet.cpp [hunk 7]
```diff
@@ -0,0 +1,121 @@
+#include "wincapture.h"
+#include "winrospublisher.h"
```
Comment (NaveenThum) [inline src/wincameranodelet.cpp:4]: Copyrights?

File: src/wincapture.cpp [hunk 8]
```diff
@@ -0,0 +1,358 @@
+#include "wincapture.h"
+using namespace winrt;
```
Comment (NaveenThum) [inline src/wincapture.cpp:3]: Copyrights?

File: src/wincameranodelet.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+            }
+            else
+            {
+                camera1.attach(new ros_win_camera::WindowsMFCapture(isDevice, winrt::to_hstring(videoSourcePath), false));
+                camera1->ChangeCaptureConfig(Width, Height, frameRate, MFVideoFormat_ARGB32);
+                camera->AddSampleHandler(handler1);
+                camera1->AddSampleHandler(handler);
+            }
+
+            //Sleep(3000);
```
Comment (ooeygui) [inline src/wincameranodelet.cpp]: Commented code.

File: src/winrospublisher.cpp [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#include "winrospublisher.h"
+namespace ros_win_camera
+{
+
```
Comment (ooeygui) [inline src/winrospublisher.cpp:5]: Is this a placeholder?

File: src/wincameranode.cpp [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+                        }
+                        else
+                        {
+                            ROS_WARN("Setting resolution  failed. Use rescale_camera_info param for rescaling");
+                        }
+                    });
+            }
+            else
+            {
+                //castd::cout << "Received SAmple\n";
```
Comment (ooeygui) [inline src/wincameranode.cpp]: Commented Code

File: src/winrospublisher.cpp [hunk 12]
```diff
@@ -0,0 +1,5 @@
+#include "winrospublisher.h"
+namespace ros_win_camera
```
Comment (NaveenThum) [inline src/winrospublisher.cpp:3]: Copyrights?

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+
+## Declare a cpp library
+add_library(win_camera STATIC src/winrospublisher.cpp src/wincapture.cpp)
+
+add_dependencies(win_camera win_camera_generate_messages_cpp)
+
+add_library(win_camera_nodelet SHARED src/wincameranodelet.cpp)
+set_target_properties(win_camera_nodelet PROPERTIES WINDOWS_EXPORT_ALL_SYMBOLS TRUE) 
+ 
+add_compile_options ("/std:c++17")
```
Comment (ooeygui) [inline CMakeLists.txt]: C++17 is used multiple times; you should only need one of these.

File: src/wincameranodelet.cpp [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            //ros::NodeHandle private_node("~");
+            auto privateNode = getPrivateNodeHandle();
+            std::string videoSourcePath = "";
+            bool isDevice = true;
+            float frameRate = 30;
+            winrt::com_ptr< ros_win_camera::WindowsMFCapture> camera, camera1;
+            std::string frame_id("camera");
+            privateNode.getParam("frame_rate", frameRate);
+
+            winrt::slim_mutex waitForFinish;
```
Comment (seanyen) [inline src/wincameranodelet.cpp]: nit: can we use standard library thread primitives here? https://github.com/microsoft/xlang/issues/476

**Review by ooeygui — CHANGES_REQUESTED** [review]
Could you add a sample launch file?
Eventually we'll want to add tests which exercise this. 
We have a CI/CD pipeline in the works.

File: src/winrospublisher.cpp (hunk 10 above)
Comment (Sudhaushu) [reply src/winrospublisher.cpp:5]: yes. I plan to move the definitions from the header to here..

---
In reply to: [383535343](https://github.com/ms-iot/ros_win_camera/pull/2#discussion_r383535343) [](ancestors = 383535343)

File: src/wincameranodelet.cpp (hunk 3 above)
Comment (Sudhaushu) [reply src/wincameranodelet.cpp]: Forgot to mention in the description... I am still working on the nodelet code. This code will change significantly in my updates to come..

---
In reply to: [383532875](https://github.com/ms-iot/ros_win_camera/pull/2#discussion_r383532875) [](ancestors = 383532875)

**Review by ooeygui — APPROVED** [review]

File: package.xml [hunk 15]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0"?>
+<package>
+  <name>win_camera</name>
+  <version>0.1.0</version>
+  <description>win_camera uses Windows Media Foundation to capture camera image.
+  </description>
+  <maintainer email="susohoni@microsoft.com">Sudhanshu Sohoni</maintainer>
+  <license>BSD</license>
+
+  <url type="website">http://wiki.ros.org/cv_camera</url>
```
Comment (seanyen) [inline package.xml]: nit: remove those `cv_camera` entries.

File: src/wincameranode.cpp [hunk 16]
```diff
@@ -0,0 +1,140 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+#include "wincapture.h"
+#include "winrospublisher.h"
+#include <ros/ros.h>
+#include <string>
+using namespace winrt::Windows::System::Threading;
+using namespace winrt::Windows::Foundation;
+using namespace ros_win_camera;
+const int32_t PUBLISHER_BUFFER_SIZE = 4;
```
Comment (seanyen) [inline src/wincameranode.cpp]: Just curious why the buffer size is hard-coded as 4 here? Is it something should be configured by the users?

File: src/wincameranode.cpp [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+        camera1->AddSampleHandler(handler);
+    }
+#ifdef TEST_SETCAMERAINFO
+    Sleep(10000);
+    auto info = spCameraInfoManager->getCameraInfo();
+    info.height = 720;
+    info.width = 400;
+    spCameraInfoManager->setCameraInfo(info);
+#endif //#ifdef TEST_SETCAMERAINFO
+    std::cout << "\nPress enter key to stop";
```
Comment (seanyen) [inline src/wincameranode.cpp]: In ROS, you can just put `ros::spin();` here.

**Comment by Sudhaushu** [comment]
> Could you add a sample launch file?
> Eventually we'll want to add tests which exercise this.
> We have a CI/CD pipeline in the works.
I shall take that up as a separate task and PR.

**Review by seanyen — APPROVED** [review]
LGTM. @Sudhaushu Let me know if you want this to be merged first, and you can work on tests and launch file support in separate PRs.

**Comment by Sudhaushu** [comment]
> LGTM. @Sudhaushu Let me know if you want this to be merged first, and you can work on tests and launch file support in separate PRs.
@seanyen - please go ahead and merge the PR. I shall start a separate PR for the tests and launch file.

**Pull request merged**